### PR TITLE
spartan: Add bind ip blacklist and forward zones

### PR DESCRIPTION
**Pull request by nlsun**
## High Level Description

Adds a blacklist to the ips that spartan will bind on. This can be used on masters to allow for something like dnsmasq to run without sacrificing other functionality.

Also adds a way to delegate zones to different resolvers.

## Related Issues

  - [DCOS-15214](https://jira.mesosphere.com/browse/DCOS-15214)
  - [DCOS-15310](https://jira.mesosphere.com/browse/DCOS-15310)

## Checklist for all PR's

  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [ ] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [ ] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

**Comment by nlsun** [comment]
@drewkerrigan @GoelDeepak please review

**Comment by nlsun** [comment]
@mesosphere-mergebot label [Ready For Review]

**Comment by GoelDeepak** [comment]
lgtm

**Comment by nlsun** [comment]
@mellenburg would you like to review

**Comment by nlsun** [comment]
@mesosphere-mergebot label [Work In Progress]

**Comment by nlsun** [comment]
Still a work in progress, but what are your thoughts @branden ?

**Comment by nlsun** [comment]
@branden ready for review

**Comment by branden** [comment]
@mesosphere-mergebot label ready for review

File: gen/calc.py [hunk 1]
```diff
@@ -519,6 +529,44 @@ def validate_exhibitor_storage_master_discovery(master_discovery, exhibitor_stor
...
+            assert isinstance(u, list), fz_err("{} not a list".format(u))
+            assert len(u) == 2, fz_err("{} not length 2".format(u))
+
+            ip = u[0]
+            port = u[1]
+            assert valid_ipv4_address(ip), fz_err("{} not a valid IP address".format(ip))
+            validate_int_in_range(port, 1, 65535)
+
+
+def fz_err(msg):
```
Comment (branden) [inline gen/calc.py]: It'd be nice to define this within the body of `validate_dns_forward_zones()`, since it's only relevant there.

**Comment by branden** [comment]
@nsun Please make sure to document the new `dns_bind_ip_blacklist` and `dns_forward_zones` config params for 1.10 in https://github.com/dcos/dcos-docs and https://github.com/mesosphere/dcos-docs-enterprise. Looks like we don't have a 1.10 copy of the docs yet, but you'll need to add these to `1.10/installing/custom/configuration-parameters.md`.

**Comment by nlsun** [comment]
I moved the fz_err() into the function itself. Also added a test to trigger the error at least once.

**Comment by nlsun** [comment]
Is the Azure test broken?
```
[22:00:09][Step 1/1] [2017-05-02 22:00:09,719|requests.packages.urllib3.connectionpool|INFO]: Starting new HTTPS connection (1): management.azure.com
[22:00:10][Step 1/1] [2017-05-02 22:00:10,096|test_util.azure|INFO]: Waiting for deployment. Current state: Running
[22:00:10][Step 1/1] Traceback (most recent call last):
[22:00:10][Step 1/1]   File "/teamcity/work/19ad0dff41587ee6/env/bin/test-azure-rm-deploy", line 11, in <module>
[22:00:10][Step 1/1]     sys.exit(main())
[22:00:10][Step 1/1]   File "/teamcity/work/19ad0dff41587ee6/env/lib/python3.5/site-packages/test_util/test_azure.py", line 149, in main
[22:00:10][Step 1/1]     dcos_resource_group.wait_for_deployment()
[22:00:10][Step 1/1]   File "/teamcity/work/19ad0dff41587ee6/env/lib/python3.5/site-packages/test_util/azure.py", line 236, in wait_for_deployment
[22:00:10][Step 1/1]     check_deployment_operations()
[22:00:10][Step 1/1]   File "/teamcity/work/19ad0dff41587ee6/env/lib/python3.5/site-packages/retrying.py", line 49, in wrapped_f
[22:00:10][Step 1/1]     return Retrying(*dargs, **dkw).call(f, *args, **kw)
[22:00:10][Step 1/1]   File "/teamcity/work/19ad0dff41587ee6/env/lib/python3.5/site-packages/retrying.py", line 214, in call
[22:00:10][Step 1/1]     raise RetryError(attempt)
[22:00:10][Step 1/1] retrying.RetryError: RetryError[Attempts: 46, Value: False]
[22:00:10][Step 1/1] Process exited with code 1
[22:00:10][Step 1/1] Process exited with code 1
[22:00:10][Step 1/1] Step Command Line failed
```

**Comment by nlsun** [comment]
Docs PR here: https://github.com/dcos/dcos-docs/pull/1043

**Comment by nlsun** [comment]
@branden Tests are passing, is this shippable?

**Comment by branden** [comment]
Tests are passing (except for a flake in the upgrade test) and we have 2 accepting reviews, so yeah this is shippable.

**Comment by branden** [comment]
@mesosphere-mergebot label ship it

**Pull request merged**